// nodebb-plugin-user-invitations/lib/plugin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.views = exports.invitations = exports.rewards = exports.conditions = exports.conditionals = exports.init = undefined;

var _nodebb = require('./nodebb');

var nodebb = _interopRequireWildcard(_nodebb);

var _fs = require('fs');

var fs = _interopRequireWildcard(_fs);

var _path = require('path');

var path = _interopRequireWildcard(_path);

var _conditionals = require('./conditionals');

var conditionals = _interopRequireWildcard(_conditionals);

var _conditions = require('./conditions');

var conditions = _interopRequireWildcard(_conditions);

var _rewards = require('./rewards');

var rewards = _interopRequireWildcard(_rewards);

var _invitations = require('./invitations');

var invitations = _interopRequireWildcard(_invitations);

var _views = require('./views');

var views = _interopRequireWildcard(_views);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function init(data, callback) {
  // Socket events for user pages.
  nodebb.socketPlugins.invitation = {
    send: invitations.socketio.userSend,
    uninvite: invitations.socketio.userUninvite,
    reinvite: invitations.socketio.userReinvite
  };

  nodebb.socketAdmin.invitation = {
    send: invitations.socketio.adminSend,
    reinvite: invitations.socketio.adminReinvite,
    giveReward: invitations.giveReward,
    setInvites: invitations.setInvites
  };

  views.init(data, callback);
}

exports.init = init;
exports.conditionals = conditionals;
exports.conditions = conditions;
exports.rewards = rewards;
exports.invitations = invitations;
exports.views = views;
