// nodebb-plugin-calendar/build/lib/event.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.escapeEvent = exports.filterByPid = exports.getEventsByDate = exports.getEvent = exports.eventExists = exports.saveEvent = exports.deleteEvent = undefined;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _responses = require('./responses');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const db = require.main.require('./src/database');
const privileges = require.main.require('./src/privileges');
const plugins = require.main.require('./src/plugins');
const topics = require.main.require('./src/topics');

const p = _bluebird2.default.promisify;

const sortedSetAdd = p(db.sortedSetAdd);
const sortedSetRemove = p(db.sortedSetRemove);
const getSortedSetRangeByScore = p(db.getSortedSetRangeByScore);
const setObject = p(db.setObject);
const getObject = p(db.getObject);
const getObjects = p(db.getObjects);
const deleteKey = p(db.delete);
const exists = p(db.exists);
const filterPids = p(privileges.posts.filter);
const fireHook = p(plugins.fireHook);
const getTopicsFields = p(topics.getTopicsFields);
const getTopicField = p(topics.getTopicField);

const listKey = 'plugins:calendar:events';

const saveEvent = event => _bluebird2.default.all([sortedSetAdd(listKey, event.startDate, `${ listKey }:pid:${ event.pid }`), setObject(`${ listKey }:pid:${ event.pid }`, event)]);

const deleteEvent = pid => _bluebird2.default.all([sortedSetRemove(listKey, `${ listKey }:pid:${ pid }`), deleteKey(`${ listKey }:pid:${ pid }`), (0, _responses.removeAll)(pid)]);

const getEventsByDate = (() => {
  var ref = (0, _bluebird.coroutine)(function* (startDate, endDate) {
    const keys = yield getSortedSetRangeByScore(listKey, 0, -1, startDate, endDate);
    const events = yield getObjects(keys);

    const topicsWithCids = yield getTopicsFields(events.map(function (ev) {
      return ev.tid;
    }), ['cid']);

    return events.map(function (ev, i) {
      return _extends({}, ev, {
        cid: topicsWithCids[i].cid
      });
    });
  });
  return function getEventsByDate(_x, _x2) {
    return ref.apply(this, arguments);
  };
})();

const eventExists = pid => exists(`${ listKey }:pid:${ pid }`);

const getEvent = (() => {
  var ref = (0, _bluebird.coroutine)(function* (pid) {
    const event = yield getObject(`${ listKey }:pid:${ pid }`);

    var _ref = yield getTopicField(event.tid, 'cid');

    const cid = _ref.cid;


    return _extends({}, event, {
      cid: cid
    });
  });
  return function getEvent(_x3) {
    return ref.apply(this, arguments);
  };
})();

const filterByPid = (events, uid) => filterPids('read', events.map(e => e.pid), uid).then(filtered => events.filter(e => filtered.includes(e.pid)));

const escapeEvent = (() => {
  var ref = (0, _bluebird.coroutine)(function* (event) {
    var _ref2 = yield _bluebird2.default.all([fireHook('filter:parse.raw', event.location), fireHook('filter:parse.raw', event.description)]);

    var _ref3 = _slicedToArray(_ref2, 2);

    const location = _ref3[0];
    const description = _ref3[1];


    return _extends({}, event, {
      location: location,
      description: description
    });
  });
  return function escapeEvent(_x4) {
    return ref.apply(this, arguments);
  };
})();

exports.deleteEvent = deleteEvent;
exports.saveEvent = saveEvent;
exports.eventExists = eventExists;
exports.getEvent = getEvent;
exports.getEventsByDate = getEventsByDate;
exports.filterByPid = filterByPid;
exports.escapeEvent = escapeEvent;
