// nodebb-plugin-calendar/build/lib/event.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.escapeEvent = exports.filterByPid = exports.getEventsByDate = exports.getEvent = exports.eventExists = exports.saveEvent = exports.deleteEvent = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _responses = require('./responses');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const db = require.main.require('./src/database');
const privileges = require.main.require('./src/privileges');
const plugins = require.main.require('./src/plugins');
const posts = require.main.require('./src/posts');

const p = _bluebird2.default.promisify;

const sortedSetAdd = p(db.sortedSetAdd);
const sortedSetRemove = p(db.sortedSetRemove);
const getSortedSetRangeByScore = p(db.getSortedSetRangeByScore);
const setObject = p(db.setObject);
const getObject = p(db.getObject);
const getObjects = p(db.getObjects);
const deleteKey = p(db.delete);
const exists = p(db.exists);
const filterPids = p(privileges.posts.filter);
const fireHook = p(plugins.fireHook);
const getCidsByPids = p(posts.getCidsByPids);
const getCidByPid = p(posts.getCidByPid);

const listKey = 'plugins:calendar:events';

const saveEvent = event => _bluebird2.default.all([sortedSetAdd(listKey, event.startDate, `${ listKey }:pid:${ event.pid }`), setObject(`${ listKey }:pid:${ event.pid }`, event)]);

const deleteEvent = pid => _bluebird2.default.all([sortedSetRemove(listKey, `${ listKey }:pid:${ pid }`), deleteKey(`${ listKey }:pid:${ pid }`), (0, _responses.removeAll)(pid)]);

const getEventsByDate = (() => {
  var ref = (0, _bluebird.coroutine)(function* (startDate, endDate) {
    const keys = yield getSortedSetRangeByScore(listKey, 0, -1, startDate, endDate);
    const events = yield getObjects(keys);

    const cids = yield getCidsByPids(events.map(function (event) {
      return event.pid;
    }));

    return events.map(function (event, i) {
      return (0, _extends3.default)({}, event, {
        cid: cids[i]
      });
    });
  });
  return function getEventsByDate(_x, _x2) {
    return ref.apply(this, arguments);
  };
})();

const eventExists = pid => exists(`${ listKey }:pid:${ pid }`);

const getEvent = (() => {
  var ref = (0, _bluebird.coroutine)(function* (pid) {
    const event = yield getObject(`${ listKey }:pid:${ pid }`);
    const cid = yield getCidByPid(event.pid);

    return (0, _extends3.default)({}, event, {
      cid: cid
    });
  });
  return function getEvent(_x3) {
    return ref.apply(this, arguments);
  };
})();

const filterByPid = (events, uid) => filterPids('read', events.map(e => e.pid), uid).then(filtered => events.filter(e => filtered.includes(e.pid)));

const escapeEvent = (() => {
  var ref = (0, _bluebird.coroutine)(function* (event) {
    var _ref = yield _bluebird2.default.all([fireHook('filter:parse.raw', event.location), fireHook('filter:parse.raw', event.description)]);

    var _ref2 = (0, _slicedToArray3.default)(_ref, 2);

    const location = _ref2[0];
    const description = _ref2[1];


    return (0, _extends3.default)({}, event, {
      location: location,
      description: description
    });
  });
  return function escapeEvent(_x4) {
    return ref.apply(this, arguments);
  };
})();

exports.deleteEvent = deleteEvent;
exports.saveEvent = saveEvent;
exports.eventExists = eventExists;
exports.getEvent = getEvent;
exports.getEventsByDate = getEventsByDate;
exports.filterByPid = filterByPid;
exports.escapeEvent = escapeEvent;
