// nodebb-plugin-calendar/build/lib/postSave.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.postEditCallback = exports.postSaveCallback = exports.postSave = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _validator = require('validator');

var _validator2 = _interopRequireDefault(_validator);

var _parse = require('./parse');

var _parse2 = _interopRequireDefault(_parse);

var _privileges = require('./privileges');

var _event = require('./event');

var _validateEvent3 = require('./validateEvent');

var _validateEvent4 = _interopRequireDefault(_validateEvent3);

var _reminders = require('./reminders');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const plugins = require.main.require('./src/plugins');
// const winston = require.main.require('winston');

const log = function log() {
  var _console;

  return (_console = console).log.apply(_console, arguments);
};
const p = _bluebird2.default.promisify;

const fireHook = p(plugins.fireHook);

const regex = new RegExp('(\\[\\s?event\\s?\\][\\w\\W]*\\[\\s?\\/\\s?event\\s?\\])|' + '(\\[\\s?event\\-invalid?\\s?\\][\\w\\W]*\\[\\s?\\/\\s?event\\-invalid?\\s?\\])');

const postSave = (() => {
  var ref = (0, _bluebird.coroutine)(function* (postData) {
    let event = (0, _parse2.default)(postData.content);

    // delete event if no longer in post
    if (!postData.content.match(regex)) {
      const existed = yield (0, _event.eventExists)(postData.pid);
      if (existed) {
        yield (0, _reminders.notify)({
          event: yield (0, _event.getEvent)(postData.pid),
          message: '[[calendar:event_deleted]]'
        });

        yield (0, _event.deleteEvent)(postData.pid);
        log(`[plugin-calendar] Event (pid:${ postData.pid }) deleted`);
      }

      return postData;
    }

    const invalid = function invalid() {
      const d = postData;
      d.content = d.content.replace(/\[\s?(\/?)\s?event\s?\]/g, '[$1event-invalid]');
      return postData;
    };

    if (!event) {
      return invalid();
    }

    var _validateEvent = (0, _validateEvent4.default)(event);

    var _validateEvent2 = (0, _slicedToArray3.default)(_validateEvent, 2);

    const failed = _validateEvent2[0];
    const failures = _validateEvent2[1];

    if (failed) {
      const obj = failures.reduce(function (val, failure) {
        return (0, _extends3.default)({}, val, {
          [failure]: event[failure]
        });
      }, {});
      log(`[plugin-calendar] Event (pid:${ postData.pid }) validation failed: `, obj);
      return invalid();
    }

    const can = yield (0, _privileges.canPostEvent)(postData.tid, postData.uid);
    if (!can) {
      return invalid();
    }

    event.name = _validator2.default.escape(event.name);
    event.location = event.location.trim();
    event.description = event.description.trim();
    event.pid = postData.pid;
    event.uid = postData.uid;
    event = yield fireHook('filter:plugin-calendar:event.post', event);

    yield (0, _event.saveEvent)(event);
    log(`[plugin-calendar] Event (pid:${ event.pid }) saved`);

    return postData;
  });
  return function postSave(_x) {
    return ref.apply(this, arguments);
  };
})();

const postSaveCallback = (postData, cb) => postSave(postData).asCallback(cb);
const postEditCallback = (data, cb) => postSave(data.post).then(() => data).asCallback(cb);

exports.postSave = postSave;
exports.postSaveCallback = postSaveCallback;
exports.postEditCallback = postEditCallback;
