// nodebb-plugin-calendar/build/lib/postSave.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.postEditCallback = exports.postSaveCallback = exports.postSave = undefined;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _validator = require('validator');

var _validator2 = _interopRequireDefault(_validator);

var _parse = require('./parse');

var _parse2 = _interopRequireDefault(_parse);

var _privileges = require('./privileges');

var _event = require('./event');

var _validateEvent3 = require('./validateEvent');

var _validateEvent4 = _interopRequireDefault(_validateEvent3);

var _reminders = require('./reminders');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const plugins = require.main.require('./src/plugins');
// const winston = require.main.require('winston');

const log = (...args) => console.log(...args);
const p = _bluebird2.default.promisify;

const fireHook = p(plugins.fireHook);

const regex = new RegExp('(\\[\\s?event\\s?\\][\\w\\W]*\\[\\s?\\/\\s?event\\s?\\])|' + '(\\[\\s?event\\-invalid?\\s?\\][\\w\\W]*\\[\\s?\\/\\s?event\\-invalid?\\s?\\])');

const postSave = (() => {
  var ref = (0, _bluebird.coroutine)(function* (postData) {
    let event = (0, _parse2.default)(postData.content);

    // delete event if no longer in post
    if (!postData.content.match(regex)) {
      const existed = yield (0, _event.eventExists)(postData.pid);
      if (existed) {
        yield (0, _reminders.notify)({
          event: yield (0, _event.getEvent)(postData.pid),
          message: '[[calendar:event_deleted]]'
        });

        yield (0, _event.deleteEvent)(postData.pid);
        log(`[plugin-calendar] Event (pid:${ postData.pid }) deleted`);
      }

      return postData;
    }

    const invalid = function invalid() {
      return _extends({}, postData, {
        content: postData.content.replace(/\[\s?(\/?)\s?event\s?\]/g, '[$1event-invalid]')
      });
    };

    if (!event) {
      return invalid();
    }

    var _validateEvent = (0, _validateEvent4.default)(event);

    var _validateEvent2 = _slicedToArray(_validateEvent, 2);

    const failed = _validateEvent2[0];
    const failures = _validateEvent2[1];

    if (failed) {
      const obj = failures.reduce(function (val, failure) {
        return _extends({}, val, {
          [failure]: event[failure]
        });
      }, {});
      log(`[plugin-calendar] Event (pid:${ postData.pid }) validation failed: `, obj);
      return invalid();
    }

    const can = yield (0, _privileges.canPostEvent)(postData.pid, postData.uid);
    if (!can) {
      return invalid();
    }

    event.name = _validator2.default.escape(event.name);
    event.pid = postData.pid;
    event.tid = postData.tid;
    event.uid = postData.uid;
    event = yield fireHook('filter:plugin-calendar:event.post', event);

    yield (0, _event.saveEvent)(event);
    log(`[plugin-calendar] Event (pid:${ event.pid }) saved`);

    return postData;
  });
  return function postSave(_x) {
    return ref.apply(this, arguments);
  };
})();

const postSaveCallback = (postData, cb) => postSave(postData).asCallback(cb);
const postEditCallback = (data, cb) => postSave(data.post).then(postData => _extends({}, data, {
  post: postData
})).asCallback(cb);

exports.postSave = postSave;
exports.postSaveCallback = postSaveCallback;
exports.postEditCallback = postEditCallback;
