// nodebb-plugin-calendar/build/lib/reminders.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.notify = exports.initNotifierDaemon = undefined;

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _responses = require('./responses');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const db = require.main.require('./src/database');
const notifications = require.main.require('./src/notifications');
const posts = require.main.require('./src/posts');
const meta = require.main.require('./src/meta');

// import { fork } from 'child_process';

const p = _bluebird2.default.promisify;

const getSortedSetRange = p(db.getSortedSetRange);
const getObjectsFields = p(db.getObjectsFields);
const createNotif = p(notifications.create);
const pushNotif = p(notifications.push);
const getPostField = p(posts.getPostField);
const getSetting = p(meta.settings.getOne);

const listKey = 'plugins:calendar:events';

const getAll = (() => {
  var ref = (0, _bluebird.coroutine)(function* () {
    const keys = yield getSortedSetRange(listKey, 0, -1);
    const events = yield getObjectsFields(keys, ['pid', 'reminders', 'startDate']);

    return events;
  });
  return function getAll() {
    return ref.apply(this, arguments);
  };
})();

const notify = (() => {
  var ref = (0, _bluebird.coroutine)(function* (_ref) {
    let event = _ref.event;
    let reminder = _ref.reminder;
    let message = _ref.message;

    let uids;
    // if reminder is for the event start
    // notify 'maybe' and 'yes' responders
    // otherwise, notify only 'yes' responders
    if (reminder === 0) {
      const responses = yield (0, _responses.getAll)({
        pid: event.pid,
        selection: ['yes', 'maybe']
      });
      uids = [].concat((0, _toConsumableArray3.default)(responses.yes), (0, _toConsumableArray3.default)(responses.maybe));
    } else {
      const responses = yield (0, _responses.getAll)({
        pid: event.pid,
        selection: ['yes']
      });
      uids = responses.yes;
    }

    const content = yield getPostField(event.pid, 'content');
    const notif = yield createNotif({
      bodyShort: `[[calendar:event_starting, ${ message }, ${ event.name }]]`,
      bodyLong: content,
      nid: `plugin-calendar:tid:${ event.tid }:pid:${ event.pid }:event`,
      pid: event.pid,
      tid: event.tid,
      from: event.uid,
      path: `/post/${ event.pid }`,
      importance: 1
    });
    yield pushNotif(notif, uids);
  });
  return function notify(_x) {
    return ref.apply(this, arguments);
  };
})();

const initNotifierDaemon = (() => {
  var ref = (0, _bluebird.coroutine)(function* () {
    // ms between checking for reminders
    // pulled from settings
    const checkingInterval = yield getSetting('plugin-calendar', 'checkingInterval');

    const checkReminders = (() => {
      var ref = (0, _bluebird.coroutine)(function* () {
        // timespan we check is a checkingInterval in the future
        // so as to avoid sending notifications too late
        const start = Date.now() + checkingInterval;
        const end = start + checkingInterval;

        const events = yield getAll();

        const mom = (0, _moment2.default)(start);

        yield _bluebird2.default.all(events.map(function (event) {
          const reminder = [0].concat((0, _toConsumableArray3.default)(event.reminders)).find(function (r) {
            const remDate = event.startDate - r;
            return remDate > start && remDate < end;
          });
          if (reminder === 0) {
            return {
              event: event,
              reminder: reminder,
              message: '[[calendar:now]]'
            };
          }
          if (reminder) {
            const message = mom.to(event.startDate);
            return { event: event, reminder: reminder, message: message };
          }
          return null;
        }).filter(Boolean).map(notify));
      });
      return function checkReminders() {
        return ref.apply(this, arguments);
      };
    })();

    const daemon = function daemon() {
      checkReminders().asCallback(function () {
        return setTimeout(daemon, checkingInterval);
      });
    };
    daemon();
  });
  return function initNotifierDaemon() {
    return ref.apply(this, arguments);
  };
})();

exports.initNotifierDaemon = initNotifierDaemon;
exports.notify = notify;
