// nodebb-plugin-calendar/build/lib/responses.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getUserResponse = exports.getAll = exports.removeAll = exports.submitResponse = undefined;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _privileges = require('./privileges');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const db = require.main.require('./src/database');
const user = require.main.require('./src/user');

const p = _bluebird2.default.promisify;

const setAdd = p(db.setAdd);
const setsRemove = p(db.setsRemove);
const deleteAll = p(db.deleteAll);
const getSetsMembers = p(db.getSetsMembers);
const isSetMember = p(db.isSetMember);
const getUsersFields = p(user.getUsersFields);

const listKey = 'plugins:calendar:events';
const values = ['yes', 'maybe', 'no'];

const submitResponse = (() => {
  var ref = (0, _bluebird.coroutine)(function* (_ref) {
    let pid = _ref.pid;
    let uid = _ref.uid;
    let value = _ref.value;

    if (!values.includes(value)) {
      throw new Error('[[error:invalid-data]]');
    }
    if (!(yield (0, _privileges.canViewPost)(pid, uid))) {
      throw new Error('[[error:no-privileges]]');
    }

    yield _bluebird2.default.all([setsRemove(values.filter(function (val) {
      return val !== value;
    }).map(function (val) {
      return `${ listKey }:pid:${ pid }:responses:${ val }`;
    }), uid), setAdd(`${ listKey }:pid:${ pid }:responses:${ value }`, uid)]);
  });
  return function submitResponse(_x) {
    return ref.apply(this, arguments);
  };
})();

const removeAll = pid => deleteAll(values.map(val => `${ listKey }:pid:${ pid }:responses:${ val }`));

const getAll = (() => {
  var ref = (0, _bluebird.coroutine)(function* () {
    var _ref2 = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    let pid = _ref2.pid;
    var _ref2$uid = _ref2.uid;
    let uid = _ref2$uid === undefined ? false : _ref2$uid;
    var _ref2$selection = _ref2.selection;
    let selection = _ref2$selection === undefined ? values : _ref2$selection;

    if (uid !== false && !(yield (0, _privileges.canViewPost)(pid, uid))) {
      throw new Error('[[error:no-privileges]]');
    }

    const responseUids = yield getSetsMembers(selection.map(function (val) {
      return `${ listKey }:pid:${ pid }:responses:${ val }`;
    }));
    const userFields = ['userslug', 'picture', 'username', 'icon:bgColor', 'icon:text'];

    var _ref3 = yield _bluebird2.default.all(responseUids.map(function (uids) {
      return getUsersFields(uids, userFields);
    }));

    var _ref4 = _slicedToArray(_ref3, 3);

    const yes = _ref4[0];
    const maybe = _ref4[1];
    const no = _ref4[2];


    return {
      yes: yes,
      maybe: maybe,
      no: no
    };
  });
  return function getAll(_x2) {
    return ref.apply(this, arguments);
  };
})();

const getUserResponse = (() => {
  var ref = (0, _bluebird.coroutine)(function* (_ref5) {
    let pid = _ref5.pid;
    let uid = _ref5.uid;

    if (!(yield (0, _privileges.canViewPost)(pid, uid))) {
      throw new Error('[[error:no-privileges]]');
    }

    const arr = yield _bluebird2.default.all(values.map(function (val) {
      return isSetMember(`${ listKey }:pid:${ pid }:responses:${ val }`, uid);
    }));

    return values[arr.findIndex(function (val) {
      return !!val;
    })];
  });
  return function getUserResponse(_x4) {
    return ref.apply(this, arguments);
  };
})();

exports.submitResponse = submitResponse;
exports.removeAll = removeAll;
exports.getAll = getAll;
exports.getUserResponse = getUserResponse;
